---
title: "R Exercices PH525.3x"
output: html_notebook
---
#WEEK 1

## R Refresher Exercises
### R Refresher Exercises #1

Download and install the GSE5859Subset package then load the data.


```{r}
library(devtools)
install_github("genomicsclass/GSE5859Subset")
library(GSE5859Subset)
data(GSE5859Subset) ##this loads the three tables
```


How many samples where processed on 2005-06-27?


```{r}
length(grep('2005-06-27', sampleInfo$date))
```

### R Refresher Exercises #2

How many of the genes represented in this particular technology are on chromosome Y?

```{r}
length(which(geneAnnotation == "chrY"))
```

### R Refresher Exercises #3

What is the log expression value of the for gene ARPC1A on the one subject that we measured on 2005-06-10?

```{r}
pid = geneAnnotation$PROBEID[which(geneAnnotation$SYMBOL== 'ARPC1A')]
samDate = sampleInfo$filename[(grep('2005-06-10', sampleInfo$date))]

geneExpression[pid,samDate]
```


### R Refresher Exercises #4

Use the function apply to find the median value of each column. What is the median value of these values?

```{r}
firstMed = apply(geneExpression, 2, median)
secMed = median(firstMed)
secMed
```

###  R Refresher Exercises #5

This problem is more advanced than the previous ones. Note that it might take you some time to solve and you should feel free to seek help in the discussion forum. The exercises is meant to motivate you to learn a an imporant R skills: creating functions to use with apply.

Write a function that takes a vector of values e and a binary vector group coding two groups, and returns the p-value from a t-test: t.test( e[group==1], e[group==0])$p.value.

Now define g to code cases (1) and controls (0) like this g <- factor(sampleInfo$group)

Next use the function apply to run a t-test for each row of geneExpression and obtain the p-value. What is smallest p-value among all these t-tests? 

```{r}

tpval = function(e, group){
  group = as.factor(group)
  pVal = t.test(e[group==1], e[group == 0])$p.value
  return(pVal)
}

e = rnorm(20)
g0 = as.factor(sample(c(0,1), length(e), replace = T))
# t.test(e[g0==1], e[g0 == 0])$pvalue

res = tpval(e, g0)
res

g = g <- factor(sampleInfo$group)

sort(apply(geneExpression,1,tpval, g))[1]


```

## Inference in Practice Exercises
### Inference in Practice Exercises #1

These exercises will help clarify that p-values are random variables and some of the properties of these p-values. Note that just like the sample average is a random variable because it is based on a random sample, the p-values are based on random variables (sample mean and sample standard deviation for example) and thus it is also a random variable.

To see this, let's see how p-values change when we take different samples.

```{r}

set.seed(1)
library(downloader)
url = "https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/femaleControlsPopulation.csv"
filename = "femaleControlsPopulation.csv"
if (!file.exists(filename)) {
  download(url,destfile=filename)
}
population = read.csv(filename)
pvals <- replicate(1000,{
  control = sample(population[,1],12)
  treatment = sample(population[,1],12)
  t.test(treatment,control)$p.val
})
head(pvals)
hist(pvals)
    
```
What proportion of the p-values is below 0.05?

```{r}
length(pvals[pvals < 0.05])/length(pvals)
```
What proportion of the p-values is below 0.05?
```{r}
length(pvals[pvals < 0.01])/length(pvals)
```


### Inference in Practice Exercises #3 

Assume you are testing the effectiveness of 20 diets on mice weight. For each of the 20 diets you run an experiment with 10 control mice and 10 treated mice. Assume the null hypothesis that the diet has no effect is true for all 20 diets and that mice weights follow a normal distribution with mean 30 grams and a standard deviation of 2 grams, run a Monte Carlo simulation for one of these studies:

```{r}
cases = rnorm(10,30,2)
controls = rnorm(10,30,2)
t.test(cases,controls)
```



Now run a Monte Carlo simulation imitating the results for the experiment for all 20 diets. If you set the seed at 100, set.seed(100), and use the same code as above inside a call to replicate how many of p-values (number not proportion) are below 0.05?

```{r}
set.seed(100)

mousePval = function(){
  cases = rnorm(10,30,2)
  controls = rnorm(10,30,2)
  return(t.test(cases,controls)$p.value)
}

pvals = c()
for (i in 1:20){
  pvals[i] = mousePval()
}

length(pvals[pvals < 0.05])
```

## Inference in Practice Exercises #4

Now create a simulation to learn about the distribution of the number of p-values that are less than 0.05. In question 1.2.3 we ran the 20 diet experiment once. Now we will run these 20 experiments 1,000 times and each time save the number of p-values that are less than 0.05.

Set the seed at 100 again, set.seed(100), run the code from Question 1.2.3 1,000 times, and save the number of times the p-value is less than 0.05 for each of the 1,000 instances.

What is the average of these 1,000 numbers? Note that this is the expected number of tests (out of the 20 we run) that we will reject when the null is true. (Hint: use replicate twice)


```{r}
set.seed(100)

lenVec = c()
mousePval = function(){
  cases = rnorm(10,30,2)
  controls = rnorm(10,30,2)
  return(t.test(cases,controls)$p.value)
}

for (j in 1:1000){
  pvals = c()
  for (i in 1:20){
    pvals[i] = mousePval()
  }

  lenVec[j] = length(pvals[pvals < 0.05])

}

mean(lenVec)

# Can also use the replicate function : replicate(n , FUN)

```

## Inference in Practice Exercises #5


Note that what the answer to question #4 says is that on average, we expect some p-value to be 0.05 even when the null is true for all diets.

Using the same simulation data from the question above, for what proportion of the 1,000 replications do we reject the null hypothesis at least once (more than 0 false positives)? (Enter your response as a decimal value -- i.e. 0.10 for 10%.)

```{r}
length(which(lenVec >= 1 ))/length(lenVec)
```


# WEEK 2

Which procedure is more conservative (picks less genes, i.e. rejects less null hypothesis): Bonferroni's or Sidak's?
Make a plot of $\alpha/m$ and $1-(1-\alpha)^{1/m}$ and for various values of m>1. 

```{r}
alpha <- seq(0,0.25,0.01)
m = 1:10000
bonferroni = alpha/m
sidak = 1-(1 - alpha)**(1/m)
plot(m, sidak, col = "red", ylab = "k", pch = 20)
points(m, bonferroni, col = "green", pch = 21)
```

### Bonferroni Correction Exercises #2 (Monte Carlo Simulation) 

Monte Carlo simulation. To simulate the p-value results of, say, 8,793 t-tests for which the null is true we don't actual have to generate the original data. As we learned in class we can generate p-values from a uniform distribution like this:

`pvals <- runif(8793,0,1)`

Using what we have learned, set the cutoff using the Bonferroni correction that guarantees an FWER lower than 0.05 and report back the FWER. Set the seed at 1, `set.seed(1)` and run 10,000 simulation. Report the Monte Carlo estimate of the FWER below.

```{r}
set.seed(1)
alpha = 0.05
m = 8793
k = alpha/m
estimate = c()
for (i in 1:10000){
  pvals <- runif(m,0,1)
  estimate[i] = sum(pvals<k)
}

mean(estimate)


```

Using the same seed repeat the above for Sidak's cutoff. 


```{r}
set.seed(1)
alpha = 0.05
m = 8793
k = 1-(1 - alpha)**(1/m)
estimate = c()
for (i in 1:10000){
  pvals <- runif(m,0,1)
  estimate[i] = sum(pvals<k)
}

mean(estimate)

```


## FDR Exercises

In this assessment we will define error controlling procedures for experimental data. We will make list of genes based on q-values. We will also assess your understanding of false positives rates and false negative rates by asking you to create a Monte Carlo simulation.

```{r}
library(qvalue)
library(devtools)
library(rafalib)
install_github("genomicsclass/GSE5859Subset")
install_bioc("genefilter")
install_bioc("qvalue")
```


### FDR Exercises #1

Load the gene expression data

```{r}
library(GSE5859Subset)
data(GSE5859Subset)
```

We are interested in comparing gene expression between the two groups defined in the sampleInfo table.

Compute a p-value for each gene using the function rowttests from the genefilter package in Bioconductor.

```{r}
library(genefilter)
?rowttests
sum(rowttests(geneExpression, as.factor(sampleInfo$group))$p.value < 0.05)
```
### FDR Exercises #2

Apply the Bonferroni correction to the p-values obtained in question #1 to achieve a FWER of 0.05. How many genes are called significant under this procedure?

```{r}
sum(rowttests(geneExpression, as.factor(sampleInfo$group))$p.value < (0.05/nrow(geneExpression))) 

```
### FDR Exercises #3

Note that the FDR is a property of a list of features, not each specific feature. The q-value relates FDR to an individual feature. To define the q-value we order features we tested by p-value then compute the FDRs for a list with the most significant, the two most significant, the three most significant, etc... The FDR of the list with the, say, m most significant tests is defined as the q-value of the m-th most significant feature. In other words, the q-value of a feature, is the FDR of the biggest list that includes that gene.

In R, we can compute the q-value using the `p.adjust` function with the FDR option. Read the help file for `p.adjust` and then, for our gene expression dataset, compute how many genes achieve an FDR < 0.05

```{r}
pvals = rowttests(geneExpression, as.factor(sampleInfo$group))$p.value
sum(p.adjust(pvals, method = "fdr", n = length(pvals))<0.05)

```


### FDR Exercises #4

Now use the qvalue function, in the Bioconductor qvalue package, to estimate q-values using the procedure described by Storey.

Using this estimate how many genes have q-values below 0.05?


```{r}
sum(qvalue(pvals)$qvalues<0.05)
```


### FDR Exercises #5

Read the help file for qvalue and report the estimated proportion of genes for which the null hypothesis is true $\pi_{0} = m_{0}/m$

```{r}
qvalue(pvals)$pi0
```

### FDR Exercises #7

Note that this is an advanced question and that you can ask questions in the discussion forum.

Create a Monte Carlo Simulation in which you simulate measurements from 8,793 genes for 24 samples: 12 cases and 12 controls. 

```{r}
n <- 24
m <- 8793
mat <- matrix(rnorm(n*m),m,n)
```

Now for 500 genes, there is a difference of 2 between cases and controls:

```{r}
delta <- 2
positives <- 500
mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
```

So the null hypothesis is true for 8793-500 genes. Using the notation from the videos m=8793, m0=8293 and m1=500

Set the seed at 1, set.seed(1) and run this experiment 1,000 times with a Monte Carlo simulation. For each instance compute p-values using a t-test (using rowttests in the genefilter package) and create three lists of genes using:

- Bonferroni correction to achieve an FWER of 0.05,

```{r}
set.seed(1)
library(qvalue)
library(genefilter)
n <- 24
m <- 8793
m0=8293
m1=500
B <- 1000
delta <-2
positives <- 500
g <- factor(rep(c(0,1),each=12))
result <- replicate(B,{
  mat <- matrix(rnorm(n*m),m,n)
  mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
  pvals = rowttests(mat,g)$p.val
  ##Bonferroni
  FP1 <- sum(pvals[-(1:positives)]<=0.05/m)  
  })
mean(result/m0)
```

- p-adjust estimates of FDR to achieve an FDR of 0.05, and
- qvalue estimates of FDR to to achieve an FDR of 0.05.

For each of these three lists compute the number of false positives in the list and the number of false negatives: genes not in the list that should have been because the null hypothesis is not true (we added 2 to the controls to create the cases).

What is the false positive rate (false positives divided by m0) if we use Bonferroni?

### FDR Exercises #8

From the same Monte Carlo simulation as in the question above, what is the false negative rate if we use Bonferroni?

```{r}
set.seed(1)
library(qvalue)
library(genefilter)
n <- 24
m <- 8793
B <- 1000
m0=8293
m1=500
delta <-2
positives <- 500
g <- factor(rep(c(0,1),each=12))
result <- replicate(B,{
  mat <- matrix(rnorm(n*m),m,n)
  mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
  pvals = rowttests(mat,g)$p.val
  ##Bonferroni
  FP1 <- sum(pvals[-(1:positives)]<=0.05/m)  
  FN1 <- sum(pvals[1:positives]>0.05/m)
  c(FP1,FN1)
})
mean(result[2,]/m1)
```

### FDR Exercises #9

From the same Monte Carlo simulation as in question #7, what is the false positive rate if we use q-values from p.adjust?
```{r}
set.seed(1)
library(qvalue)
library(genefilter)
n <- 24
m <- 8793
m0=8293
m1=500
B <- 1000
delta <-2
positives <- 500
g <- factor(rep(c(0,1),each=12))
result <- replicate(B,{
  mat <- matrix(rnorm(n*m),m,n)
  mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
  pvals = rowttests(mat,g)$p.val
  padj = p.adjust(pvals, method = "fdr", n = m)
  FP1 <- sum(padj[-(1:positives)]<=0.05)
  })

mean(result/m0)
```

 FDR Exercises #10

From the same Monte Carlo simulation as in question #7, what is the false negative rate if we use q-values from p.adjust?

```{r}
set.seed(1)
library(qvalue)
library(genefilter)
n <- 24
m <- 8793
m0=8293
m1=500
B <- 1000
delta <-2
positives <- 500
g <- factor(rep(c(0,1),each=12))
result <- replicate(B,{
  mat <- matrix(rnorm(n*m),m,n)
  mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
  pvals = rowttests(mat,g)$p.val
  padj = p.adjust(pvals, method = "fdr", n = m)
  FP1 <- sum(padj[-(1:positives)]<=0.05)
  FN1 <- sum(padj[1:positives]>0.05)
  c(FP1,FN1)
  })

mean(result[2,]/m1)
```


### FDR Exercises #11

From the same Monte Carlo simulation as in question #7, what is the false positive rate if we use q-values from qvalue function?

```{r}
set.seed(1)
library(qvalue)
library(genefilter)
n <- 24
m <- 8793
m0=8293
m1=500
B <- 1000
delta <-2
positives <- 500
g <- factor(rep(c(0,1),each=12))
result <- replicate(B,{
  mat <- matrix(rnorm(n*m),m,n)
  mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
  pvals = rowttests(mat,g)$p.val
  padj = qvalue(pvals)$qvalues
  FP1 <- sum(padj[-(1:positives)]<=0.05)
  FN1 <- sum(padj[1:positives]>0.05)
  c(FP1,FN1)
  })

mean(result[1,]/m0)
```


### FDR Exercises #12


From the same Monte Carlo simulation as in question #7, what is the false negative rate if we use q-values from qvalue function?

```{r}
set.seed(1)
library(qvalue)
library(genefilter)
n <- 24
m <- 8793
m0=8293
m1=500
B <- 1000
delta <-2
positives <- 500
g <- factor(rep(c(0,1),each=12))
result <- replicate(B,{
  mat <- matrix(rnorm(n*m),m,n)
  mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
  pvals = rowttests(mat,g)$p.val
  padj = qvalue(pvals)$qvalues
  FP1 <- sum(padj[-(1:positives)]<=0.05)
  FN1 <- sum(padj[1:positives]>0.05)
  c(FP1,FN1)
  })

mean(result[2,]/m1)
```