---
title: "R Exercices PH525.3x"
output: html_notebook
---
#WEEK 1

## R Refresher Exercises
### R Refresher Exercises #1

Download and install the GSE5859Subset package then load the data.


```{r}
library(devtools)
install_github("genomicsclass/GSE5859Subset")
library(GSE5859Subset)
data(GSE5859Subset) ##this loads the three tables
```


How many samples where processed on 2005-06-27?


```{r}
length(grep('2005-06-27', sampleInfo$date))
```

### R Refresher Exercises #2

How many of the genes represented in this particular technology are on chromosome Y?

```{r}
length(which(geneAnnotation == "chrY"))
```

### R Refresher Exercises #3

What is the log expression value of the for gene ARPC1A on the one subject that we measured on 2005-06-10?

```{r}
pid = geneAnnotation$PROBEID[which(geneAnnotation$SYMBOL== 'ARPC1A')]
samDate = sampleInfo$filename[(grep('2005-06-10', sampleInfo$date))]

geneExpression[pid,samDate]
```


### R Refresher Exercises #4

Use the function apply to find the median value of each column. What is the median value of these values?

```{r}
firstMed = apply(geneExpression, 2, median)
secMed = median(firstMed)
secMed
```

###  R Refresher Exercises #5

This problem is more advanced than the previous ones. Note that it might take you some time to solve and you should feel free to seek help in the discussion forum. The exercises is meant to motivate you to learn a an imporant R skills: creating functions to use with apply.

Write a function that takes a vector of values e and a binary vector group coding two groups, and returns the p-value from a t-test: t.test( e[group==1], e[group==0])$p.value.

Now define g to code cases (1) and controls (0) like this g <- factor(sampleInfo$group)

Next use the function apply to run a t-test for each row of geneExpression and obtain the p-value. What is smallest p-value among all these t-tests? 

```{r}

tpval = function(e, group){
  group = as.factor(group)
  pVal = t.test(e[group==1], e[group == 0])$p.value
  return(pVal)
}

e = rnorm(20)
g0 = as.factor(sample(c(0,1), length(e), replace = T))
# t.test(e[g0==1], e[g0 == 0])$pvalue

res = tpval(e, g0)
res

g = g <- factor(sampleInfo$group)

sort(apply(geneExpression,1,tpval, g))[1]


```

## Inference in Practice Exercises
### Inference in Practice Exercises #1

These exercises will help clarify that p-values are random variables and some of the properties of these p-values. Note that just like the sample average is a random variable because it is based on a random sample, the p-values are based on random variables (sample mean and sample standard deviation for example) and thus it is also a random variable.

To see this, let's see how p-values change when we take different samples.

```{r}

set.seed(1)
library(downloader)
url = "https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/femaleControlsPopulation.csv"
filename = "femaleControlsPopulation.csv"
if (!file.exists(filename)) {
  download(url,destfile=filename)
}
population = read.csv(filename)
pvals <- replicate(1000,{
  control = sample(population[,1],12)
  treatment = sample(population[,1],12)
  t.test(treatment,control)$p.val
})
head(pvals)
hist(pvals)
    
```
What proportion of the p-values is below 0.05?

```{r}
length(pvals[pvals < 0.05])/length(pvals)
```
What proportion of the p-values is below 0.05?
```{r}
length(pvals[pvals < 0.01])/length(pvals)
```


### Inference in Practice Exercises #3 

Assume you are testing the effectiveness of 20 diets on mice weight. For each of the 20 diets you run an experiment with 10 control mice and 10 treated mice. Assume the null hypothesis that the diet has no effect is true for all 20 diets and that mice weights follow a normal distribution with mean 30 grams and a standard deviation of 2 grams, run a Monte Carlo simulation for one of these studies:

```{r}
cases = rnorm(10,30,2)
controls = rnorm(10,30,2)
t.test(cases,controls)
```



Now run a Monte Carlo simulation imitating the results for the experiment for all 20 diets. If you set the seed at 100, set.seed(100), and use the same code as above inside a call to replicate how many of p-values (number not proportion) are below 0.05?

```{r}
set.seed(100)

mousePval = function(){
  cases = rnorm(10,30,2)
  controls = rnorm(10,30,2)
  return(t.test(cases,controls)$p.value)
}

pvals = c()
for (i in 1:20){
  pvals[i] = mousePval()
}

length(pvals[pvals < 0.05])
```

## Inference in Practice Exercises #4

Now create a simulation to learn about the distribution of the number of p-values that are less than 0.05. In question 1.2.3 we ran the 20 diet experiment once. Now we will run these 20 experiments 1,000 times and each time save the number of p-values that are less than 0.05.

Set the seed at 100 again, set.seed(100), run the code from Question 1.2.3 1,000 times, and save the number of times the p-value is less than 0.05 for each of the 1,000 instances.

What is the average of these 1,000 numbers? Note that this is the expected number of tests (out of the 20 we run) that we will reject when the null is true. (Hint: use replicate twice)


```{r}
set.seed(100)

lenVec = c()
mousePval = function(){
  cases = rnorm(10,30,2)
  controls = rnorm(10,30,2)
  return(t.test(cases,controls)$p.value)
}

for (j in 1:1000){
  pvals = c()
  for (i in 1:20){
    pvals[i] = mousePval()
  }

  lenVec[j] = length(pvals[pvals < 0.05])

}

mean(lenVec)

# Can also use the replicate function : replicate(n , FUN)

```

## Inference in Practice Exercises #5


Note that what the answer to question #4 says is that on average, we expect some p-value to be 0.05 even when the null is true for all diets.

Using the same simulation data from the question above, for what proportion of the 1,000 replications do we reject the null hypothesis at least once (more than 0 false positives)? (Enter your response as a decimal value -- i.e. 0.10 for 10%.)

```{r}
length(which(lenVec >= 1 ))/length(lenVec)
```


# WEEK 2

Which procedure is more conservative (picks less genes, i.e. rejects less null hypothesis): Bonferroni's or Sidak's?
Make a plot of $\alpha/m$ and $1-(1-\alpha)^{1/m}$ and for various values of m>1. 

```{r}
alpha <- seq(0,0.25,0.01)
m = 1:10000
bonferroni = alpha/m
sidak = 1-(1 - alpha)**(1/m)
plot(m, sidak, col = "red", ylab = "k", pch = 20)
points(m, bonferroni, col = "green", pch = 21)
```

### Bonferroni Correction Exercises #2 (Monte Carlo Simulation) 

Monte Carlo simulation. To simulate the p-value results of, say, 8,793 t-tests for which the null is true we don't actual have to generate the original data. As we learned in class we can generate p-values from a uniform distribution like this:

`pvals <- runif(8793,0,1)`

Using what we have learned, set the cutoff using the Bonferroni correction that guarantees an FWER lower than 0.05 and report back the FWER. Set the seed at 1, `set.seed(1)` and run 10,000 simulation. Report the Monte Carlo estimate of the FWER below.

```{r}
set.seed(1)
alpha = 0.05
m = 8793
k = alpha/m
estimate = c()
for (i in 1:10000){
  pvals <- runif(m,0,1)
  estimate[i] = sum(pvals<k)
}

mean(estimate)


```

Using the same seed repeat the above for Sidak's cutoff. 


```{r}
set.seed(1)
alpha = 0.05
m = 8793
k = 1-(1 - alpha)**(1/m)
estimate = c()
for (i in 1:10000){
  pvals <- runif(m,0,1)
  estimate[i] = sum(pvals<k)
}

mean(estimate)

```

